src/coreclr/jit/morph.cpp

```diff
@@ -3922,8 +3936,12 @@ GenTree* Compiler::fgMorphMultiregStructArg(CallArg* arg)
                 newArg->AddField(this, argIndir, offset, argIndir->TypeGet());
             }
 
+#ifndef TARGET_LOONGARCH64
             // Make sure we loaded exactly the required amount of bytes.
+            // But for LoongArch64's ABI, the struct {long a; float b;} may be passed
+            // by Integer register and float register while must be considered the padding here.
             assert(structSize == (lastElem.Offset + genTypeSize(lastElem.Type)));
+#endif
```

Comment:
This is not right, this path must always load the exact size, it means the logic for setting up `elems` is not correct.

What does the `elems` array look like for the cited struct?

Reply:
structSize=16; for struct {long a; float b;}

Reply:
<pre>
3941	            assert(structSize == (lastElem.Offset + genTypeSize(lastElem.Type)));
(gdb) p structSize
$1 = 16
(gdb) p lastElem
$2 = (Compiler::ArgElem &) @0xfffffee970: {Type = TYP_FLOAT, Offset = 8}
</pre>

Reply:
And the second element is 4 bytes in size, I see.

Only some subset of structs can pass 4 bytes only instead of the full 8 (i. e. with the padding). For example, with this struct:
```cs
[StructLayout(LayoutKind.Sequential, Size = 16)]
struct StructWithSingnificantPadding
{
    long L;
    float F;
}
```
All 16 bytes must be passed (we assume the user could store something in the explicit padding).

It seems simpler to always pass all the bytes (this would also be consistent with Unix x64).

Reply:
Thanks very much for your reviewing and advices!

Reply:
Well, the problem is not really the assert, it's the behavior for structs like `StructWithSignificantPadding`, right.

Reply:
> Well, the problem is not really the assert, it's the behavior for structs like `StructWithSignificantPadding`, right.

Yes, you are right.
But for LA's ABI, it should be loaded as single float for the field `float F`.

Reply:
> Why is that?
> 
> Reading the ABI, it does allow undefined padding bits:
> 
> ```
> Bits unused due to padding, and bits past the end of a structure whose size in bits is not divisible by GRLEN, are undefined
> ```
> 
> Note `StructWithSignificantPadding` is effectively outside the native ABI, since we put a stronger guarantee on "the padding" for it.

The struct {double a;  int b;}, should loading the second field `int b` should be loaded as int32 and should not be loaded int64 directly.
Because the high 32bits must be sign-extend.

What should I do now?

Reply:
> The struct {double a; int b;}, should loading the second field int b should be loaded as int32 and should not be loaded int64 directly. Because the high 32bits must be sign-extend.

Hmm, that is a very unfortunate requirement. Could you please point me to the part of the specification which implies this (for future references)? Does it apply only to "last" fields? On the surface of things it seems to contradict the "padding is undefined" statement.

With this requirement, there is a need to distinguish between `Size = 16` (we'll need a custom ABI for such cases, effectively), and the "natural" padding.

> What should I do now?

I would recommend opening an issue (bug) about this problem and linking to it from this code.

Reply:
> And the second element is 4 bytes in size, I see.
> 
> Only some subset of structs can pass 4 bytes only instead of the full 8 (i. e. with the padding). For example, with this struct:
> 
> ```cs
> [StructLayout(LayoutKind.Sequential, Size = 16)]
> struct StructWithSignificantPadding
> {
>     long L;
>     float F;
>     [padding: 4]
> }
> ```
> 
> All 16 bytes must be passed (we assume the user could store something in the explicit padding).
> 
> It seems simpler to always pass all the bytes (this would also be consistent with Unix x64).

I feel uncertain about this padding here.
why must be loaded the padding as part of the struct liking it is a extra field.

Reply:
If I understand it clearly, I will try to modify the LA's implement.

Reply:
>  why must be loaded the padding as part of the struct liking it is a extra field.

It's a choice we make. Essentially, for some structs we assume the user can manipulate them using pointers/ref arithmetic and read/store something in the padding.

The canonical example of such "struct with significant padding" is a fixed-sized buffer as implemented by the C# compiler. It has one field of the array's element type and then `Size = CountOfElements * sizeof(ElementType)`. See [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQhArgHYDOAhgGYwAE5ALlMWI3QGICWCMAJgGUuALxgAhYtVpRCAb0J1FdDAGY61HvzrAAno3oBBKFEo6A2kgC6AbkIBfIA==).

See also #71711.

Reply:
thanks very much.
I will reply later, maybe tomorrow night.

Reply:
> Only some subset of structs can pass 4 bytes only instead of the full 8 (i. e. with the padding). For example, with this struct:
> 
> ```cs
> [StructLayout(LayoutKind.Sequential, Size = 16)]
> struct StructWithSignificantPadding
> {
>     long L;
>     float F;
>     [padding: 4]
> }
> ```
> 
> All 16 bytes must be passed (we assume the user could store something in the explicit padding).

Or the struct
```cs
[StructLayout(LayoutKind.Sequential, Size = 16)]
struct StructWithSignificantPadding
{
    double d;
    int i;
    [padding: 4]
}
```

* (1) all 16 bytes mus be passed.
For the CPUs liking the LA64 and MIPS64, the struct passed to the callee function, if the struct's field within the register which including the padding does't be stored on the stack within the prolog and directly as an operand for a operation, liking this,
```
### the register A1 containg the second field with padding
IG_01:
       ###prolog, and A1 not be stored to stack
IG_02:
       ###the 32bits operation liking  add_w, compare_w and branch:
       addi.w    T0, A1, 0x4        ### the hight 32bits of the A1 must be sign extend, or the result is unpredictable.
                                               ### This is quite different with the AArch64 and AMD64.
```
* (2) I want to know the padding will be used at this case ? 
   For example, the register which including the padding will be passed to an instruction directly liking the previous example ？
* (3) I want more detail info about the usage of the padding to amend the LA64's ABI implementation,
  for example, adding more details info about the padding and specially treat it before using it directly, maybe store the padding first and then sign extend the hight 32bits.

Reply:
So essentially the question is how should the "custom ABI" for these structs look like. Not a simple problem.

One thing to consider is that we must retain compatibility with the native ABI for C/C++ PInvokes.

For example, the following struct:
```cs
[StructLayout(LayoutKind.Explicit)]
struct S
{
    [FieldOffset(0)]
    public long U1;
    [FieldOffset(0)]
    public long U2;
    [FieldOffset(8)]
    public int I;
}
```
Equivalent to this native one:
```c
struct S
{
    union
    {
        long long U1;
        long long U2;
    }
    int I;
}
```
Would be classified as having "significant padding" in the managed world. So it would have to be passed one way to managed methods and another way to native ones.

> For the CPUs liking the LA64 and MIPS64, the struct passed to the callee function, if the struct's field within the register which including the padding doesn't be stored on the stack

Structs with significant padding will not be promoted, and so the situation you cite ought not to arise (in the Jitted code at least).

The general question does still stand on how the padding should be treated though (with a "it is sequence of `byte` fields"-like rule), so we can reason about it.

It looks like a significant decision to make, so I will ask @jkotas and @dotnet/jit-contrib for assistance here.

My personal thinking here is that we should make this as simple as possible, perhaps going as far as passing all structs with significant padding by reference. I suppose this also requires fixing #71711 first.

Reply:
I do not have an opinion on this one. LA64 ABI for passing structs is extremely complex. I do not have all details internalized.

> One thing to consider is that we must retain compatibility with the native ABI for C/C++ PInvokes.
> For example, the following struct:

Yes, native ABI compatibility is a good thing to shoot for. However, we do not have 100% interop for unions today. I believe that there are cases on existing platforms where it is not possible to define a struct with unions in a portable way in C#.

Reply:
> Yes, native ABI compatibility is a good thing to shoot for. 
Agree with you.

> However, we do not have 100% interop for unions today. I believe that there are cases on existing platforms where it is not possible to define a struct with unions in a portable way in C#.
Yes, I think so that defining a struct with union is not a good way for keeping portable.